frontend/ui/widget/scrolltextwidget.lua

```diff
@@ -66,24 +72,15 @@ function ScrollTextWidget:init()
     end
 end
 
-function ScrollTextWidget:updateScrollBar(text)
-    local virtual_line_num = text:getVirtualLineNum()
-    local visible_line_count = text:getVisLineCount()
-    local all_line_count = text:getAllLineCount()
-    self.v_scroll_bar:set(
-        (virtual_line_num - 1) / all_line_count,
-        (virtual_line_num - 1 + visible_line_count) / all_line_count
-    )
-end
-
 function ScrollTextWidget:scrollText(direction)
     if direction == 0 then return end
     if direction > 0 then
-        self.text_widget:scrollDown()
+        low, high = self.text_widget:scrollDown()
+        self.v_scroll_bar:set(low, high)
     else
-        self.text_widget:scrollUp()
+        low, high = self.text_widget:scrollUp()
+        self.v_scroll_bar:set(low, high)
```

Comment:
Since it's the same in both branches, it's better to move `self.v_scroll_bar:set(low, high)` outside of the if condition.